### PR TITLE
Avoid possible test timeout due to very verbose output

### PR DESCRIPTION
**Pull request by dsplaisted**
The `diag` case of `ItUsesVerbosityPassedToDefineVerbosityOfConsoleLoggerOfTheTests` runs `dotnet test -v diag`, which makes MSBuild emit hundreds of thousands of lines of output. `TestCommand.Execute` echoes every captured line to the xUnit `ITestOutputHelper`; xUnit buffers it and flushes hundreds of megabytes over the test host's single IPC channel when the test completes.  That can saturates the channel, and cause xUnit to think the test didn't finish and cancel it after 15 minutes.

This adds a new `TruncatingTestOutputHelper` that forwards only a bounded head and tail of the output (by character budget) to the log and drops the middle. This keeps useful diagnostics from both ends while preventing the IPC flood, and preserves the diagnostic-verbosity coverage.

Observed failure: https://dev.azure.com/dnceng-public/public/_build/results?buildId=1453617 on https://github.com/dotnet/sdk/pull/54592

### CODE REVIEW

File: test/Microsoft.NET.TestFramework/TruncatingTestOutputHelper.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.NET.TestFramework
+{
+    /// <summary>
```
Comment (MichaelSimons) [inline test/Microsoft.NET.TestFramework/TruncatingTestOutputHelper.cs:6]: Suggestion: The tail eviction behavior (dropping entire lines to stay within budget) could be called out more explicitly in the doc comments — right now you have to figure this out from the implementation.

File: test/Microsoft.NET.TestFramework/TruncatingTestOutputHelper.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                _omittedCharacters += message.Length - _maxTailCharacters;
+                message = message.Substring(message.Length - _maxTailCharacters);
+            }
+
+            _tail.Enqueue(message);
+            _tailCharacters += message.Length;
+
+            while (_tailCharacters > _maxTailCharacters && _tail.Count > 1)
+            {
+                string dropped = _tail.Dequeue();
```
Comment (MichaelSimons) [inline test/Microsoft.NET.TestFramework/TruncatingTestOutputHelper.cs:101]: Concern: Because eviction operates on whole lines, there's an edge case where the tail ends up with essentially no useful content. For example: suppose the second-to-last line is 60,000 characters and the final line is very brief. The 60K line gets evicted to stay under the 50K budget, leaving only the whitespace line in the tail buffer. In that scenario the "most recent 50,000 characters" guarantee is effectively hollow — you get almost nothing. Is that acceptable, or should there be a fallback (e.g., substring the oversized line from the end) to ensure the tail always contains meaningful content?

**Review by nagilson — APPROVED** [review]

**Review by joeloff — APPROVED** [review]
Is the idea that we add this as new issues show up in tests?

**Comment by MichaelSimons** [comment]
@dsplaisted - Just wanted to circle back on [my review comments in this PR](https://github.com/dotnet/sdk/pull/54642#pullrequestreview-4486952422). I may have missed a response, but I didn't see any follow-up on them.

Could you take a look when you get a chance? Even if you disagree with the feedback or don't plan to make changes, I'd appreciate a response so we can close the loop on the review.

**Comment by dsplaisted** [comment]
@MichaelSimons I saw those last week and agreed with them, but this morning I saw the PR was green and didn't remember the comments before merging.

Right now the truncating log is used on a single test, so I don't think it's critical to improve how it handles the truncation.  It may be a good idea to expand this to be used in all tests by default, at which point we should make the improvements.

**Comment by MichaelSimons** [comment]
> Right now the truncating log is used on a single test, so I don't think it's critical to improve how it handles the truncation. It may be a good idea to expand this to be used in all tests by default, at which point we should make the improvements.

That's fair, the misleading comments was the more concerning point.
